Source/engine/random.hpp

```diff
@@ -136,6 +137,130 @@ class DiabloGenerator {
 	}
 };
 
+// Based on fmix32 implementation from MurmurHash3 created by Austin Appleby in 2008
+// https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash3.cpp#L68
+// and adapted from https://prng.di.unimi.it/splitmix64.c written in 2015 by Sebastiano Vigna
+//
+// See also:
+//  Guy L. Steele, Doug Lea, and Christine H. Flood. 2014.
+//  Fast splittable pseudorandom number generators. SIGPLAN Not. 49, 10 (October 2014), 453–472.
+//  https://doi.org/10.1145/2714064.2660195
+class SplitMix32 {
+	uint32_t state;
+
+public:
+	SplitMix32(uint32_t state)
+	    : state(state)
+	{
+	}
+
+	uint32_t next()
+	{
+		uint32_t z = (state += 0x9e3779b9);
+		z = (z ^ (z >> 16)) * 0x85ebca6b;
+		z = (z ^ (z >> 13)) * 0xc2b2ae35;
+		return z ^ (z >> 16);
+	}
+
+	void generate(uint32_t *begin, const uint32_t *end)
+	{
+		while (begin != end) {
+			*begin = next();
+			++begin;
+		}
+	}
+};
+
+// Adapted from https://prng.di.unimi.it/splitmix64.c written in 2015 by Sebastiano Vigna
+//
+// See also:
+//  Guy L. Steele, Doug Lea, and Christine H. Flood. 2014.
+//  Fast splittable pseudorandom number generators. SIGPLAN Not. 49, 10 (October 2014), 453–472.
+//  https://doi.org/10.1145/2714064.2660195
+class SplitMix64 {
+	uint64_t state;
+
+public:
+	SplitMix64(uint64_t state)
+	    : state(state)
+	{
+	}
+
+	uint64_t next()
+	{
+		uint64_t z = (state += 0x9e3779b97f4a7c15);
+		z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
+		z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
+		return z ^ (z >> 31);
+	}
+
+	void generate(uint64_t *begin, const uint64_t *end)
+	{
+		while (begin != end) {
+			*begin = next();
+			++begin;
+		}
+	}
+};
+
+/** Adapted from https://prng.di.unimi.it/xoshiro128plusplus.c written in 2019 by David Blackman and Sebastiano Vigna */
+class xoshiro128plusplus {
+private:
+	uint32_t s[4];
+
+public:
+	xoshiro128plusplus() { seed(static_cast<uint64_t>(1)); }
+	xoshiro128plusplus(uint32_t initialSeed) { seed(initialSeed); }
+	xoshiro128plusplus(uint64_t initialSeed) { seed(initialSeed); }
+
+	void seed(uint64_t value)
+	{
+		uint64_t seeds[2];
+		SplitMix64 seedSequence { value };
+		seedSequence.generate(seeds, seeds + 2);
+
+		s[0] = static_cast<uint32_t>(seeds[0] >> 32);
+		s[1] = static_cast<uint32_t>(seeds[0]);
+		s[2] = static_cast<uint32_t>(seeds[1] >> 32);
+		s[3] = static_cast<uint32_t>(seeds[1]);
+	}
+
+	void seed(uint32_t value)
+	{
+		SplitMix32 seedSequence { value };
+		seedSequence.generate(s, s + 4);
+	}
+
+	uint32_t next()
+	{
+		const uint32_t result = std::rotl(s[0] + s[3], 7) + s[0];
+
+		const uint32_t t = s[1] << 9;
+
+		s[2] ^= s[0];
+		s[3] ^= s[1];
+		s[1] ^= s[2];
+		s[0] ^= s[3];
+
+		s[2] ^= t;
+
+		s[3] = std::rotl(s[3], 11);
+
+		return result;
+	}
+};
+
+/**
+ * @brief Overwrite the state of the global seed generator with state derived from the given seed
+ * @param seed Seed from which to derive the new state
+ */
+void ResetSeedGenerator(uint64_t seed);
```

Comment:
That's true, I agree. I did this thinking that I'd want to have a global seed generator that is deterministic based on the game seed, but I'm starting to wonder if that's really necessary. We'll need to create new generators for dungeon generation regardless, and therefore I don't know of anything else the game seed really needs to be used for besides dungeon seed generation. Even regenerating the town seed doesn't really need to use the same source of entropy as the game itself.

---

One note about `std::random_device`. I thought about using it for game seeds, but I shied away based on the following.
https://en.cppreference.com/w/cpp/numeric/random/random_device

> `std::random_device` may be implemented in terms of an implementation-defined pseudo-random number engine if a non-deterministic source (e.g. a hardware device) is not available to the implementation. In this case each `std::random_device` object may generate the same number sequence.

This may not be a realistic concern, but we do support running DevX on a bunch of fairly exotic systems. Since we got away with using `time(NULL)` for so long without players noticing any major issues, I figure our requirements aren't so stringent that we need to worry about the quality of time as an initial source of entropy for the RNG.

Reply:
Sure, I shouldn't have implied that random_device was the only choice or necessarily correct to use :D. Maybe cribbing from https://stackoverflow.com/questions/39288595/why-not-just-use-random-device/ and using a combination of the system clock and random device to generate 128bits would be enough?

> We'll need to create new generators for dungeon generation regardless, and therefore I don't know of anything else the game seed really needs to be used for besides dungeon seed generation. Even regenerating the town seed doesn't really need to use the same source of entropy as the game itself.

I figure you are referring to object/monster/item placement as well as level layout when you're talking about dungeon generation, and yeah as far as I can recall that should be enough to get to the same starting state for multiplayer games since quests state/availability is synced as part of the initial delta.

We could use the global rng more in singleplayer games (loot seeds, monster ai, damage rolls) but multiplayer games need to use the same sequences across clients to keep things in sync.

Reply:
> ... using a combination of the system clock and random device to generate 128bits would be enough?

I took some inspiration from all this spitballing. See how you like it.

> I figure you are referring to object/monster/item placement as well as level layout when you're talking about dungeon generation, and yeah as far as I can recall that should be enough to get to the same starting state for multiplayer games since quests state/availability is synced as part of the initial delta.

In this particular case, I was only really thinking about seed generation. However, it's true that monster/object/item placement is done in the same place as seed generation and also does not have to be consistent with vanilla. We could replace the LCG with xoshiro for all of it, and it should still be fine. Level layout does have to be consistent with vanilla in SP at least so I was planning to use a `DiabloGenerator` for that part.